src/map/status.cpp

```diff
@@ -10173,28 +10212,81 @@ int32 status_change_start(struct block_list* src, struct block_list* bl,enum sc_
-		case SC_VACUUM_EXTREME:
-			if (sc && sc->getSCE(SC_VACUUM_EXTREME_POSTDELAY) && sc->getSCE(SC_VACUUM_EXTREME_POSTDELAY)->val2 == val2) // Ignore post delay from other vacuum (this will make stack effect enabled)
-				return 0;
-			break;
 		case SC_STONE:
 		case SC_STONEWAIT:
 		case SC_FREEZE:
 			// Undead are immune to Freeze/Stone
-			if (undead_flag && !(flag&SCSTART_NOAVOID))
+			if (battle_check_undead(status->race, status->def_ele) != 0 && !(flag&SCSTART_NOAVOID))
 				return 0;
+			else if (type == SC_STONEWAIT) {
+				// Stonewait has a unique handling where the delay is actually duration until stone kicks in
+				val3 = max(1, tick - delay); // Petrify time
+				tick = delay;
+				delay = 0;
+			}
 			break;
 		case SC_BURNING:
 			// Level 2 Fire Element is immune
 			if (status->def_ele == ELE_FIRE && status->ele_lv == 2)
 				return 0;
 			break;
+	}
+
+	// If there is no delay, we proceed immediately
+	// Otherwise, we store the status change data in a struct and set up a timer for after the delay
+	if (delay <= 0) {
+		return status_change_start_post_delay(src, bl, type, val1, val2, val3, val4, tick, flag);
+	}
+	else {
+		delay_status* dat = ers_alloc(delay_status_ers, struct delay_status);
+		dat->src_id = src->id;
+		dat->bl_id = bl->id;
+		dat->type = type;
+		dat->val1 = val1;
+		dat->val2 = val2;
+		dat->val3 = val3;
+		dat->val4 = val4;
+		dat->tick = tick;
+		dat->flag = flag;
+
+		add_timer(gettick() + delay, status_change_start_timer, 0, (intptr_t)dat);
+	}
+
+	// Assume success
+	return 1;
+}
+
+/**
+ * Applies given status change and executes all related actions
+ * This function handles all immunities that should be checked at the end of delay
+ * @param src: Source of the status change [PC|MOB|HOM|MER|ELEM|NPC]
+ * @param bl: Target of the status change (See: enum sc_type)
+ * @param type: Status change (SC_*)
+ * @param val1~4: Depends on type of status change
+ * @param tick: Final duration of the status change
+ * @param flag: Value which determines what parts to calculate. See e_status_change_start_flags
+ * @return Whether the status change was resisted (0) or applied (1)
+ */
+int32 status_change_start_post_delay(block_list* src, block_list* bl, sc_type type, int32 val1, int32 val2, int32 val3, int32 val4, int32 tick, unsigned char flag)
+{
+	map_session_data* sd = BL_CAST(BL_PC, bl);
+	view_data* vd = status_get_viewdata(bl);
+	status_change* sc = status_get_sc(bl);
+	status_change_entry* sce;
+	status_data* status = status_get_status_data(*bl);
+	int32 undead_flag = battle_check_undead(status->race,status->def_ele);
+	int32 tick_time = 0;
+	bool sc_isnew = true;
+	std::shared_ptr<s_status_change_db> scdb = status_db.find(type);
+
+	// Check for immunities / sc fails
+	switch (type) {
+		case SC_VACUUM_EXTREME:
+			if (sc && sc->getSCE(SC_VACUUM_EXTREME_POSTDELAY) && sc->getSCE(SC_VACUUM_EXTREME_POSTDELAY)->val2 == val2) // Ignore post delay from other vacuum (this will make stack effect enabled)
```

Comment:
```suggestion
			if (sc != nullptr && sc->hasSCE(SC_VACUUM_EXTREME_POSTDELAY) && sc->getSCE(SC_VACUUM_EXTREME_POSTDELAY)->val2 == val2) // Ignore post delay from other vacuum (this will make stack effect enabled)
```